activesupport/lib/active_support/message_encryptor.rb

```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
+        cipher.encrypt
+        cipher.key = content_encryption_key
+        iv = cipher.random_iv
+        cipher.iv = iv
+        cipher.auth_data = to_ascii encode(protected_header)
+
+        ciphertext = cipher.update(@serializer.dump value) + cipher.final
+        auth_tag = cipher.auth_tag
+
+        ([protected_header, jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }).join('.')
+      end
 
-    def _encrypt(value)
-      cipher = new_cipher
-      cipher.encrypt
-      cipher.key = @secret
+      def _decrypt(encrypted_message)
+        protected_header, _encrypted_key, iv, ciphertext, auth_tag = encrypted_message.split('.').map { |a| decode a }
+        valid_header?(@serializer.load protected_header)
 
-      # Rely on OpenSSL for the initialization vector
-      iv = cipher.random_iv
+        cipher.decrypt
+        cipher.key = decrypt_content_encryption_key
+        cipher.iv = iv
+        cipher.auth_tag = auth_tag
+        cipher.auth_data = to_ascii encode(protected_header)
 
-      encrypted_data = cipher.update(@serializer.dump(value))
-      encrypted_data << cipher.final
+        decrypted_message = cipher.update(ciphertext) + cipher.final
 
-      "#{::Base64.strict_encode64 encrypted_data}--#{::Base64.strict_encode64 iv}"
-    end
+        @serializer.load decrypted_message
 
-    def _decrypt(encrypted_message)
-      cipher = new_cipher
-      encrypted_data, iv = encrypted_message.split("--").map {|v| ::Base64.strict_decode64(v)}
+      rescue OpenSSLCipherError, TypeError, ArgumentError
+        raise InvalidMessage
+      end
 
-      cipher.decrypt
-      cipher.key = @secret
-      cipher.iv  = iv
+      def cipher
+        @cipher ||= OpenSSL::Cipher.new(cipher_name)
+      end
 
-      decrypted_data = cipher.update(encrypted_data)
-      decrypted_data << cipher.final
+      def valid_header?(head)
+        raise InvalidMessage unless head.stringify_keys == header
+      end
 
-      @serializer.load(decrypted_data)
-    rescue OpenSSLCipherError, TypeError, ArgumentError
-      raise InvalidMessage
-    end
+      def header
+        @header ||= { 'typ' => 'JWE', 'alg' => @alg.to_s, 'enc' => @enc.to_s }.tap do |header|
+          header['kid'] = @options[:kid] if @options[:kid]
+        end
+      end
 
-    def new_cipher
-      OpenSSL::Cipher::Cipher.new(@cipher)
-    end
+      def to_ascii(value)
+        value.split('').map(&:ord).to_s
+      end
 
-    def verifier
-      @verifier
-    end
+      def encode(data)
+        ::Base64.strict_encode64(data)
+      end
+
+      def decode(data)
+        ::Base64.strict_decode64(data)
+      end
+
+      def content_encryption_key
+        @secret
+      end
+
+      def jwe_encrypted_key
+        if @alg == 'dir'
+          ''
+        else
+          raise UnexpectedAlgorithm, 'Unknown Encryption Algorithm.'
+        end
+      end
+
+      def decrypt_content_encryption_key
+        if @alg == 'dir'
```

Comment:
Why can people pass in `alg: 'something_else'` to `new`, but then not set it to anything? Just scrap this code and the other parts like it.